qiskit_experiments/data_processing/nodes.py

```diff
@@ -56,47 +58,216 @@ def _format_data(self, datum: Any) -> Any:
+        """
+        return np.average(datum, axis=0), np.std(datum, axis=0) / np.sqrt(datum.shape[0])
+
+
+class SVDAvg(IQPart):
+    """Singular Value Decomposition of averaged IQ data."""
+
+    def __init__(self, validate: bool = True):
+        """
+        Args:
+            validate: If set to False the DataAction will not validate its input.
+        """
+        super().__init__(validate=validate)
+        self._main_axes = None
+        self._means = None
+        self._scales = None
+
+    def _required_dimension(self) -> int:
+        """Require memory to be a 2D array."""
+        return 2
+
+    @property
+    def axis(self) -> List[np.array]:
+        """Return the axis of the trained SVD"""
+        if self._main_axes:
+            return self._main_axes
+
+        raise DataProcessorError("SVD is not trained.")
+
+    @property
+    def scales(self) -> List[float]:
+        """Return the scaling of the SVD."""
+        if self._scales:
+            return self._scales
+
+        raise DataProcessorError("SVD is not trained.")
+
+    @property
+    def is_trained(self) -> bool:
+        """Return True is the SVD has been trained.
+
+        Returns:
+            True if the SVD has been trained.
+        """
+        return self._main_axes is not None
+
+    def _process(
+        self, datum: np.array, error: Optional[np.array] = None, **options
+    ) -> Tuple[np.array, np.array]:
+        """Project the IQ data onto the axis defined by an SVD and scale it.
+
+        Args:
+            datum: A 2D array of qubits, and an average complex IQ point as [real, imaginary].
+            error: An optional 2D array of qubits, and an error on an average complex IQ
+                point as [real, imaginary].
+
+        Returns:
+            A Tuple of 1D arrays of the result of the SVD and the associated error. Each entry
+            is the real part of the averaged IQ data of a qubit.
+
+        Raises:
+            DataProcessorError: If the SVD has not been previously trained on data.
+        """
+
+        if not self.is_trained:
+            raise DataProcessorError("SVD must be trained on data before it can be used.")
+
+        n_qubits = datum.shape[0]
+        processed_data = []
+
+        if error is not None:
+            processed_error = []
+        else:
+            processed_error = None
+
+        # process each averaged IQ point with its own axis.
+        for idx in range(n_qubits):
+
+            centered = np.array([datum[idx][iq] - self._means[idx][iq] for iq in [0, 1]])
+
+            processed_data.append((self._main_axes[idx] @ centered) / self._scales[idx])
+
+            if error is not None:
+                angle = np.arctan(self._main_axes[idx][1] / self._main_axes[idx][0])
+                error_value = np.sqrt(
+                    (error[idx][0] * np.cos(angle)) ** 2 + (error[idx][1] * np.sin(angle)) ** 2
+                )
```

Comment:
is this equivalent to:
```python
np.abs(self._main_axes[idx] @ error[idx])
```
?

Reply:
Oh, but `self._main_axes[idx] @ error[idx]` is 2D. Why not carry the data through as complex so that simple expressions like this will work?

Reply:
It is done this way as this is how Qiskit terra represents the memory.